src/agent/onefuzz-agent/src/tasks/coverage/generic.rs

```diff
@@ -0,0 +1,344 @@
+            .task_id(&self.config.common.task_id);
+
+        let mut cmd = Command::new(&self.config.target_exe);
+
+        let target_options = expand.evaluate(&self.config.target_options)?;
+        cmd.args(target_options);
+
+        for (k, v) in &self.config.target_env {
+            cmd.env(k, expand.evaluate_value(v)?);
+        }
+
+        cmd.env_remove("RUST_LOG");
+        cmd.stdin(Stdio::null());
+        cmd.stdout(Stdio::piped());
+        cmd.stderr(Stdio::piped());
+
+        Ok(cmd)
+    }
+
+    pub async fn record_corpus(&mut self, dir: &Path) -> Result<usize> {
+        use futures::stream::StreamExt;
+
+        let mut corpus = fs::read_dir(dir)
+            .await
+            .map(ReadDirStream::new)
+            .with_context(|| format!("unable to read corpus directory: {}", dir.display()))?;
+
+        let mut count = 0;
+
+        while let Some(entry) = corpus.next().await {
+            match entry {
+                Ok(entry) => {
+                    if entry.file_type().await?.is_file() {
+                        self.record_input(&entry.path()).await?;
+                        count += 1;
+                    } else {
+                        warn!("skipping non-file dir entry: {}", entry.path().display());
+                    }
+                }
+                Err(err) => {
+                    error!("{:?}", err);
+                }
+            }
+        }
+
+        Ok(count)
+    }
+
+    pub async fn report_coverage_stats(&self) -> Result<()> {
+        use onefuzz_telemetry::{Event::coverage_data, EventData};
+        use EventData::*;
+
+        let s = CoverageStats::new(&self.coverage);
+        event!(coverage_data; Covered = s.covered, Features = s.features, Rate = s.rate);
+
+        Ok(())
+    }
+
+    pub async fn save_and_sync_coverage(&self) -> Result<()> {
+        let path = self.config.coverage.local_path.join(COVERAGE_FILE);
+        let text = serde_json::to_string(&self.coverage).context("serializing coverage to JSON")?;
+
+        fs::write(&path, &text)
+            .await
+            .with_context(|| format!("writing coverage to {}", path.display()))?;
+        self.config.coverage.sync_push().await?;
+
+        Ok(())
+    }
+}
+
+struct Recorded {
+    pub cache: ModuleCache,
+    pub coverage: CommandBlockCov,
+}
+
+#[cfg(target_os = "linux")]
+fn record_os_impl(mut cache: ModuleCache, cmd: Command, filter: CmdFilter) -> Result<Recorded> {
+    use coverage::block::linux::Recorder;
+
+    let mut recorder = Recorder::new(&mut cache, filter);
+    recorder.record(cmd)?;
+    let coverage = recorder.into_coverage();
+
+    Ok(Recorded { cache, coverage })
+}
+
+#[cfg(target_os = "windows")]
+fn record_os_impl(mut cache: ModuleCache, cmd: Command, filter: CmdFilter) -> Result<Recorded> {
+    use coverage::block::windows::{Recorder, RecorderEventHandler};
+
+    let mut recorder = Recorder::new(&mut cache, filter);
+    let timeout = std::time::Duration::from_secs(5);
```

Comment:
We should 1. make this timeout configurable and 2, support it similarly on Linux.